File: mikogo/unmaintained/ver0807/monitor.go
package main

// import  "parts.go"
import (
	"io/ioutil"
	"os/exec"
	"time"
	// "fmt"
)

type Monitor struct{}

// var must be Capital Letter start to Jsonfy
type infoType struct {
	CpuInfo, MemInfo, DiskInfo, Network, Uptime, Distro, Host, Version string
}

type cpuInfoType struct {
	Name, Core, Loadavg, Temperature, Idles, Totals string
}

type memInfoType struct {
	MemTotal, MemFree, Cached, Buffers, SwapTotal, SwapFree, SwapCached string
}

type diskInfoType struct {
	Usage, Detail string
}

type networkType struct {
	Dev, RxBytes, TxBytes string
}

var cpuInfo cpuInfoType
var memInfo memInfoType
var diskInfo diskInfoType
var network []networkType
var uptime, distro, host, version string

func execShell(str string) (string, string) {
	var stdout, err = exec.Command("sh", "-c", str).Output()
	if err == nil {
		return string(stdout), ""
	}
	return "", err.Error()
}

func timeHuman(time int64) string {
	var result = ""
	var timeS = int64(float64(time) + 0.5)
	var timeM = int64(float64(timeS) / 60)
	var timeH = int64(float64(timeM) / 60)
	var timeD = int64(float64(timeH) / 24)
	timeS = timeS - timeM*60
	timeM = timeM - timeH*60
	timeH = timeH - timeD*24
	result += itoa(int64(timeD)) + "d " + itoa(int64(timeH)) + "h "
	result += itoa(int64(timeM)) + "m " + itoa(int64(timeS)) + "s "
	return result
}

func getCpu() {
	if results, err := ioutil.ReadFile("/sys/class/thermal/thermal_zone0/temp"); err == nil {
		cpuInfo.Temperature = itoa(atoi(string(results)) / 1000)
	}
	if results, err := ioutil.ReadFile("/proc/loadavg"); err == nil {
		var load = strSplit(string(results), " ")
		cpuInfo.Loadavg = load[0] + " " + load[1] + " " + load[2]
	}
	if results, err := ioutil.ReadFile("/proc/cpuinfo"); err == nil {
		var info = strSplit(string(results), "\n")
		var cpuName = ""
		var cpuCore = 0
		for _, item := range info {
			if strIndex(item, "model name") == 0 {
				var tmp = strTrim(strSlice(item, strIndex(item, ":")+1, 0), " ")
				cpuCore++
				if strIndex(cpuName, tmp) == -1 {
					if cpuName != "" {
						cpuName += " | "
					}
					cpuName += tmp
				}
			}
		}
		cpuInfo.Name = cpuName
		cpuInfo.Core = itoa(int64(cpuCore))
	}
	if results, err := ioutil.ReadFile("/proc/stat"); err == nil {
		var tmpTotals, tmpIdles []string  
		var info = strSplit(string(results), "\n")
		for i:=0; i < len(info); i++ {
			if (strIndex(info[i], "cpu") == -1) {
				break;
			}
	        // cpu, cpu0, cpu1, ...
			var tmp = strSlice(info[i], strIndex(info[i], " "), 0)
			var tmpInfo = strSplit(strTrim(tmp, " "), " ")
			var tmpTotal int64 = 0
			tmpIdles = append(tmpIdles, tmpInfo[3])
			for i := 0; i < len(tmpInfo); i++ {
				tmpTotal += atoi(tmpInfo[i])
			}
			tmpTotals = append(tmpTotals, itoa(tmpTotal))
		}
		cpuInfo.Idles = jsonify(tmpIdles)
		cpuInfo.Totals = jsonify(tmpTotals)
	}
}

func getMem() {
	if results, err := ioutil.ReadFile("/proc/meminfo"); err == nil {
		var info = strSplit(string(results), "\n")
		for i := 0; i < len(info); i++ {
			var tmpInfo = info[i]
			if strIndex(tmpInfo, "MemTotal") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.MemTotal = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
			if strIndex(tmpInfo, "MemFree") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.MemFree = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
			if strIndex(tmpInfo, "Cached") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.Cached = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
			if strIndex(tmpInfo, "Buffers") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.Buffers = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
			if strIndex(tmpInfo, "SwapTotal") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.SwapTotal = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
			if strIndex(tmpInfo, "SwapFree") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.SwapFree = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
			if strIndex(tmpInfo, "SwapCached") == 0 {
				var tmp = strTrim(strSlice(tmpInfo, strIndex(tmpInfo, ":")+1, 0), " ")
				memInfo.SwapCached = strTrim(strSlice(tmp, 0, strIndex(tmp, " ")), " ")
			}
		}
	}
}

func getDisk() {
	var stdout, err = execShell("df -h 2>&1")
	if err == "" {
		diskInfo.Detail = stdout
		var info = strSplit(stdout, "\n")
		for i := 0; i < len(info); i++ {
			var tmp = strTrim(info[i], " ")
			if len(tmp) != 0 && strLastIndex(tmp, "/")+1 == len(tmp) {
				tmp = strSlice(tmp, 0, strLastIndex(tmp, "%"))
				diskInfo.Usage = strSlice(tmp, strLastIndex(tmp, " ")+1, 0)
				break
			}
		}
	}
}

func getNet() {
	if results, err := ioutil.ReadFile("/proc/net/dev"); err == nil {
		network = []networkType{}
		var info = strSplit(string(results), "\n")
		for i := 2; i < len(info); i++ {
			var tmp = strTrim(info[i], " ")
			if (tmp != "") {
				var tmpArr = strSplit(tmp, " ");
				network = append(network, networkType{strSlice(tmpArr[0], 0, -1), tmpArr[1], tmpArr[9]});
			}
		}
	}
}

func getUptime() {
	if results, err := ioutil.ReadFile("/proc/uptime"); err == nil {
		var info = strTrim(string(results), " ")
		info = strSplit(info, " ")[0]
		info = strSlice(info, 0, strIndex(info, "."))
		uptime = timeHuman(int64(atoi(info)))
	}
}

func getDistro() {
	if results, err := ioutil.ReadFile("/etc/issue"); err == nil {
		distro = string(results)
		distro = strSlice(distro, 0, strIndex(distro, "\\"))
	}
}

func getHost() {
	if results, err := ioutil.ReadFile("/etc/hostname"); err == nil {
		host = string(results)
	}
}

func getVersion() {
	if results, err := ioutil.ReadFile("/proc/version"); err == nil {
		version = string(results)
	}
}

var cTime int64 = 0
var minAskTime int64 = 3

func (this Monitor) getInfo() string {
	var tTime = time.Now().Unix() // seconds
	if tTime-cTime >= minAskTime {
		cTime = tTime
		getCpu()
		getMem()
		getDisk()
		getNet()
		getUptime()
		getDistro()
		getHost()
		getVersion()
	}
	var info infoType
	info.CpuInfo = jsonify(cpuInfo)
	info.MemInfo = jsonify(memInfo)
	info.DiskInfo = jsonify(diskInfo)
	info.Network = jsonify(network)
	info.Uptime = uptime
	info.Distro = distro
	info.Host = host
	info.Version = version
	return jsonify(info)
}

File: mikogo/unmaintained/nodejs/monitor.js
/*
 * use fs, exec, (except for those as parameters)
 * written by Mzero for MikoSite, MIT License
 * return system info = {cpuInfo, memInfo, diskInfo, process, network, uptime, distro, host, version}  
 */

var fs = require('fs');
var exec = require('child_process').exec;

module.exports = function monitor() {

    // local vars ==============================================//
    var cpuInfo = {
        "Name": "", "Core": "", "Loadavg": "",
        "Temperature": "", "Flags": "", "Idle": "", "Total": "", "Idles": "", "Totals": ""
    };
    var memInfo = {
        "MemTotal": "", "MemFree": "", "Cached": "", "Buffers": "",
        "SwapTotal": "", "SwapFree": "", "SwapCached": ""
    };
    var diskInfo = { "Usage": "", "Detail": "" };
    var process = { "General": "", "Detail": "" };
    var network = []; // { "Dev":"","Ipv4":"","Ipv6":"","RxBytes":"","TxBytes":"" }
    var uptime = "", distro = "", host = "", version = "";

    // actions ==================================================//
    getCpu(); getMem(); getDisk(); getNet();
    getUptime(); getDistro(); getHost(); getVersion();

    // local functions =========================================//
    function timeHuman(time) {
        var result = "";
        var time_s = Math.floor(time + 0.5);
        var time_m = Math.floor(time_s / 60);
        var time_h = Math.floor(time_m / 60);
        var time_d = Math.floor(time_h / 24);
        time_s = time_s - time_m * 60;
        time_m = time_m - time_h * 60;
        time_h = time_h - time_d * 24;
        result += time_d.toString() + 'd ' + time_h.toString() + 'h ';
        result += time_m.toString() + 'm ' + time_s.toString() + 's ';
        return result;
    }

    function getCpu() {
        fs.readFile("/sys/class/thermal/thermal_zone0/temp", function (err, data) {
            if (!err) cpuInfo.Temperature = String(Number(data.toString()) / 1000);
        });
        fs.readFile("/proc/loadavg", function (err, data) {
            var load = data.toString().trim().split(" ");
            cpuInfo.Loadavg = load[0] + " " + load[1] + " " + load[2];
        });
        fs.readFile("/proc/cpuinfo", function (err, data) {
            if (!err) {
                var info = data.toString().split('\n');
                var cpuName = ""; var cpuCore = 0; var cpuFlags = "";
                for (var i in info) {
                    var item = info[i];
                    if (item.indexOf("model name") == 0) {
                        var tmp = item.slice(item.indexOf(":") + 1).trim();
                        cpuCore++;
                        if (cpuName.indexOf(tmp) == -1) {
                            if (cpuName != "") {
                                cpuName += " | ";
                            }
                            cpuName += tmp;
                        }
                    }
                }
                cpuInfo.Name = cpuName;
                cpuInfo.Core = cpuCore.toString();
            }
        });
        fs.readFile("/proc/stat", function (err, data) {
            if (!err) {
                var tmpIdles = [], tmpTotals = [];
                var info = data.toString().split('\n');
                info = info.filter(function (s) { return s && s.trim(); });
                for (var i = 0; i < info.length; i++) {
                    if (info[i].indexOf('cpu') == -1) break;
                    // cpu, cpu0, cpu1, ...
                    var tmp = info[i].slice(info[i].indexOf(" "));
                    tmpInfo = tmp.trim().split(' ');
                    tmpIdles.push(Number(tmpInfo[3]));
                    var tmpTotal = 0;
                    for (var j = 0; j < tmpInfo.length; j++)
                        tmpTotal = tmpTotal + Number(tmpInfo[j]);
                    tmpTotals.push(tmpTotal);
                }
                cpuInfo.Idles = JSON.stringify(tmpIdles);
                cpuInfo.Totals = JSON.stringify(tmpTotals);
            }
        });
    }

    function getMem() {
        fs.readFile("/proc/meminfo", function (err, data) {
            if (!err) {
                var info = data.toString().split('\n');
                for (var i = 0; i < info.length; i++) {
                    var tmpInfo = info[i];
                    if (tmpInfo.indexOf("MemTotal") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.MemTotal = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                    if (tmpInfo.indexOf("MemFree") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.MemFree = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                    if (tmpInfo.indexOf("Cached") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.Cached = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                    if (tmpInfo.indexOf("Buffers") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.Buffers = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                    if (tmpInfo.indexOf("SwapTotal") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.SwapTotal = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                    if (tmpInfo.indexOf("SwapFree") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.SwapFree = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                    if (tmpInfo.indexOf("SwapCached") == 0) {
                        var tmp = tmpInfo.slice(tmpInfo.indexOf(":") + 1).trim();
                        memInfo.SwapCached = tmp.slice(0, tmp.indexOf(" ")).trim();
                    }
                }
            }
        });
    }

    function getDisk() {
        exec("df -h 2>&1", function (err, stdout, stderr) {
            if (!err) {
                diskInfo.Detail = stdout;
                var info = stdout.split('\n').filter(function (s) { return s && s.trim(); });
                for (var i in info) {
                    var tmp = info[i].trim();
                    if (tmp.length != 0 && tmp.lastIndexOf("/") + 1 == tmp.length) {
                        tmp = tmp.slice(0, tmp.lastIndexOf("%"));
                        diskInfo.Usage = tmp.slice(tmp.lastIndexOf(" "));
                        break;
                    }
                }
            }
        });
    }

    // wasted
    function getProcOri() {
        exec("ps -aux 2>&1", function (err, stdout, stderr) {
            if (!err) {
                process.Detail = stdout;
                var info = stdout.split('\n');
                var totalProc = 0;
                var users = []; var nums = [];
                for (var i = 1; i < info.length; i++) {
                    var user = info[i].trim();
                    user = user.slice(0, user.indexOf(' '));
                    if (user == "") {
                        continue
                    }
                    totalProc++;
                    var j = 0;
                    for (j = 0; j < users.length; j++) {
                        if (users[j] == user) {
                            nums[j]++;
                            break;
                        }
                    }
                    if (j == users.length) {
                        users.push(user);
                        nums.push(1);
                    }
                }
                var generalProc = "Total: " + totalProc.toString();
                for (var i = 0; i < users.length; i++) {
                    generalProc += " | " + users[i] + ": " + nums[i].toString();
                }
                process.General = generalProc;
            }
        });
    }

    // wasted 
    function getNetOri() {
        exec("ifconfig 2>&1", function (err, stdout, stderr) {
            if (!err) {
                network = [];
                var info = stdout.split('\n\n');
                for (var i in info) { // devices
                    var infoNet = {
                        "Dev": "", "Ipv4": "", "Ipv6": "", "RxBytes": "", "TxBytes": ""
                    }
                    infoNet.Dev = info[i].slice(0, info[i].indexOf(":"));
                    if (infoNet.Dev == "lo" || info[i] == "") continue;
                    var subInfo = info[i].split('\n');
                    for (var j in subInfo) {
                        var tmpInfo = subInfo[j];
                        if (tmpInfo.indexOf('inet ') != -1) {
                            var tmp = tmpInfo.slice(tmpInfo.indexOf("inet ") + 5).trim();
                            infoNet.Ipv4 = tmp.slice(0, tmp.indexOf(" ")).trim();
                        }
                        if (tmpInfo.indexOf('inet6 ') != -1) {
                            var tmp = tmpInfo.slice(tmpInfo.indexOf("inet6 ") + 6).trim();
                            infoNet.Ipv6 = tmp.slice(0, tmp.indexOf(" ")).trim();
                        }
                        if (tmpInfo.indexOf('RX packets ') != -1) {
                            var tmp = tmpInfo.slice(tmpInfo.indexOf("bytes ") + 6).trim();
                            infoNet.RxBytes = tmp.slice(0, tmp.indexOf(" ")).trim();
                        }
                        if (tmpInfo.indexOf('TX packets ') != -1) {
                            var tmp = tmpInfo.slice(tmpInfo.indexOf("bytes ") + 6).trim();
                            infoNet.TxBytes = tmp.slice(0, tmp.indexOf(" ")).trim();
                        }
                    }
                    network.push(infoNet);
                }
            }
        });
        // version 2
        fs.readFile("/proc/net/route", function (err, data) {
            if (!err) {
                var route = data.toString().split('\n').slice(1);
                var routeLength = route.length;
                for (var i = 0; i < routeLength; i++) {
                    var r4 = route[i].trim();
                    if (r4 != "") {
                        var r4Arr = r4.split("\t").filter(function (item) { return item.trim() != "" });
                        var ipv4Addr = "";
                        for (var j = 0; j < 4; j++) {
                            ipv4Addr = "." + parseInt(r4Arr[1].slice(2 * j, 2 * j + 2), 16).toString() + ipv4Addr;
                        }
                        for (var j = 0; j < network.length; j++) {
                            if (network[j].Dev == r4Arr[0]) network[j].Ipv4 += ipv4Addr.slice(1) + " | ";
                        }
                    }
                }
            }
        });
        fs.readFile("/proc/net/if_inet6", function (err, data) {
            if (!err) {
                var route6 = data.toString().split('\n');
                var route6Length = route6.length;
                for (var i = 0; i < route6Length; i++) {
                    var r6 = route6[i].trim();
                    if (r6 != "") {
                        var r6Arr = r6.split(" ").filter(function (item) { return item.trim() != "" });
                        var ipv6Addr = "";
                        for (var j = 0; j < 8; j++) {
                            ipv6Addr += ":" + r6Arr[0].slice(j * 4, j * 4 + 4);
                        }
                        for (var j = 0; j < network.length; j++) {
                            if (network[j].Dev == r6Arr[r6Arr.length - 1]) network[j].Ipv6 += ipv6Addr.slice(1) + " | ";
                        }
                    }
                }
            }
        });
    }

    function getNet() {
        fs.readFile("/proc/net/dev", function (err, data) {
            if (!err) {
                network = [];
                var dev = data.toString().split('\n').slice(2);
                var devLength = dev.length;
                for (var i = 0; i < devLength; i++) {
                    var d = dev[i].trim();
                    if (d != "") {
                        dArr = d.split(" ").filter(function (item) { return item.trim() != "" });
                        var tmp = { "Dev": "", "RxBytes": "", "TxBytes": "" };
                        tmp.Dev = dArr[0].slice(0, -1); tmp.RxBytes = dArr[1]; tmp.TxBytes = dArr[9];
                        network.push(tmp);
                    }
                }
            }
        });
    }

    function getUptime() {
        fs.readFile("/proc/uptime", function (err, data) {
            if (!err) {
                var info = data.toString().split(' ').filter(function (s) { return s && s.trim(); });
                uptime = timeHuman(Number(info[0]));
            }
        });
    }

    function getDistro() {
        fs.readFile("/etc/issue", function (err, data) {
            if (!err) {
                distro = data.toString();
                distro = distro.slice(0, distro.indexOf("\\"));
            }
        });
    }

    function getHost() {
        fs.readFile("/etc/hostname", function (err, data) {
            if (!err) { host = data.toString(); }
        });
    }

    function getVersion() {
        fs.readFile("/proc/version", function (err, data) {
            if (!err) { version = data.toString(); }
        });
    }

    // Methods ======================================================//
    var cTime = 0, minAskTime = 3000;
    this.getInfo = function () {
        // if ask twice in 3 seconds, return same value 
        var tTime = new Date().getTime();
        if (tTime - cTime > minAskTime) {
            cTime = tTime;
            getUptime(); getDistro(); getHost(); getVersion();
            getCpu(); getMem(); getDisk(); getNet();
        }
        return JSON.stringify({
            "CpuInfo": JSON.stringify(cpuInfo),
            "MemInfo": JSON.stringify(memInfo),
            "DiskInfo": JSON.stringify(diskInfo),
            "Process": JSON.stringify(process),
            "Network": JSON.stringify(network),
            "Uptime": uptime,
            "Distro": distro,
            "Host": host,
            "Version": version
        });
    }
}
